### PR TITLE
Ping init

### PR DESCRIPTION
**Pull request by annzenkina**
In the current implementation before emitting `:connected` the library does `select db`, which should be proceeded after the connection is established. Although currently selecting a database is the way how we establish if it's possible to connect to Redis. And if there are no db specified, this block will not be executed.
In this PR we execute `ping` before selecting the database, and if we don't receive pong as a reply means that the connection can't be established.

More information: https://pusher2.atlassian.net/browse/CHAN-487

### CODE REVIEW

File: spec/client_server_spec.rb [hunk 1]
```diff
@@ -62,7 +76,7 @@ def recording_server(replies = {})
 
     it 'should emit :disconnected when the connection disconnects' do
       recording_server { |server|
-        client = EM::Hiredis::Client.new('redis://localhost:6381/9')
+        client = EM::Hiredis::Client.new('redis://127.0.0.1:6381/9')
```
Comment (leesio) [inline spec/client_server_spec.rb]: I don't think these should be necessary and it makes the review harder to follow. Any chance we could remove them?
Happy to pair to try and get them working on your machine if that's the problem

File: lib/em-hiredis/redis_client.rb [hunk 2]
```diff
@@ -313,15 +313,21 @@ def factory_connection
 
         connection.on(:connected) {
           maybe_auth(connection).callback {
-            maybe_select(connection).callback {
-              @command_queue.each { |command_df, command, args|
-                connection.send_command(command_df, command, args)
+            connection.ping.callback {
```
Comment (leesio) [inline lib/em-hiredis/redis_client.rb]: Out of interest, do you happen to know whether you need to be authorised to send a `ping`? I would assume so but I don't know for sure.

Comment (annzenkina) [reply lib/em-hiredis/redis_client.rb]: yes, you need to be authorised to send a command (`ping` in this case).

File: lib/em-hiredis/pubsub_connection.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@ module EventMachine::Hiredis
   module PubsubConnection
     include EventMachine::Hiredis::EventEmitter
 
-    PUBSUB_COMMANDS = %w{subscribe unsubscribe psubscribe punsubscribe}.freeze
+    PUBSUB_COMMANDS = %w{ping subscribe unsubscribe psubscribe punsubscribe}.freeze
```
[the remark this answers is not in the document]
Comment (annzenkina) [reply lib/em-hiredis/pubsub_connection.rb]: This is not necessary, I forgot to clean it up after the first implementation where it was necessary.

File: lib/em-hiredis/redis_client.rb (hunk 2 above)
Comment (annzenkina) [reply lib/em-hiredis/redis_client.rb]: I do. I tried locally to run redis with auth, and tried to send ping before `auth` – it was not possible. To send any command to redis you need to be authorised.

File: spec/client_server_spec.rb (hunk 1 above)
Comment (annzenkina) [reply spec/client_server_spec.rb]: It doesn't work for me locally with `localhost`. As far as I know it's a known bug with MacOS. I am happy to pair if we can resolve this!

File: lib/em-hiredis/pubsub_client.rb [hunk 4]
```diff
@@ -184,7 +184,7 @@ def factory_connection
 
         connection.on(:connected) {
           maybe_auth(connection).callback {
-
+            connection.ping_with_pubsub
```
[the remark this answers is not in the document]
Comment (annzenkina) [reply lib/em-hiredis/pubsub_client.rb]: Happy do discuss that in person

**Comment by leesio** [comment]
I'm happy with this PR assuming you've tested it etc.

File: lib/em-hiredis/pubsub_client.rb [hunk 5]
```diff
@@ -181,33 +181,36 @@ def factory_connection
...
-            [ :message,
-              :pmessage,
-              :subscribe,
-              :unsubscribe,
-              :psubscribe,
-              :punsubscribe
-            ].each do |command|
-              connection.on(command) { |*args|
-                emit(command, *args)
+            connection.ping.timeout(@inactivity_response_timeout).callback {
```
Comment (annzenkina) [inline lib/em-hiredis/pubsub_client.rb]: not sure if it's a good idea to use `@inactivity_response_timeout` or shall we just use smth like `timeout(2)`

File: lib/em-hiredis/pubsub_connection.rb [hunk 6]
```diff
@@ -21,6 +20,7 @@ def initialize(inactivity_trigger_secs = nil,
         EM::Hiredis.logger.debug("#{@name} - Sending ping")
         send_command('subscribe', PING_CHANNEL)
         send_command('unsubscribe', PING_CHANNEL)
+        # send_command('ping')
```
Comment (annzenkina) [inline lib/em-hiredis/pubsub_connection.rb]: I wanted to implement this as well instead of subscribe / unsubscribe as it's nicer but didn't finish

**Comment by annzenkina** [comment]
I checked the new implementation with 1. latest version of proxy without redis running (and with redis running) 2. the broken version of proxy – seems fine. But i tried it with the broken redis - it works fine in the 2nd case, but when redis is up - it doesn't work 😅 and the `em-hiredis` in master also doesn't work with the broken proxy but when the redis is up. I couldn't figure out why, but the reason seems different to the flow we thought of.

File: lib/em-hiredis/pubsub_client.rb [hunk 7]
```diff
@@ -94,6 +94,7 @@ def initialize(
     # Commands may be issued before or during connection, they will be queued
     # and submitted to the server once the connection is active.
     def connect
+      @inactivity_response_timeout = 2 if @inactivity_response_timeout.nil?
```
Comment (annzenkina) [inline lib/em-hiredis/pubsub_client.rb]: setting the default if it was not configured.

File: spec/client_server_spec.rb [hunk 8]
```diff
@@ -31,25 +38,39 @@ def recording_server(replies = {})
       }
     end
 
-    it 'should issue select command before succeeding connection' do
+    it 'should issue ping command before succeeding connection' do
```
Comment (leesio) [inline spec/client_server_spec.rb]: This test is actually checking for `ping` and `select` 👍

File: spec/client_server_spec.rb [hunk 9]
```diff
@@ -31,25 +38,39 @@ def recording_server(replies = {})
...
+          server.connection_count.should == 1
+          server.received[0].should == 'ping'
+          done
+        }.errback { |e|
+          fail(e)
+        }
+      }
+    end
+
+    it 'should issue pinf command before emitting :connected' do
```
Comment (leesio) [inline spec/client_server_spec.rb]:
```suggestion
    it 'should issue ping command before emitting :connected' do
```

File: spec/client_server_spec.rb (hunk 8 above)
Comment (annzenkina) [reply spec/client_server_spec.rb]: true! i changed the description

File: lib/em-hiredis/pubsub_client.rb (hunk 5 above)
Comment (leesio) [reply lib/em-hiredis/pubsub_client.rb]: I think this is fine. I would maybe change the name to something like `@ping_response_timeout` though?

Comment (WillSewell) [reply lib/em-hiredis/pubsub_client.rb]: I think `inactivity_response_timeout` is fine as it's consistent with the previous name.

File: lib/em-hiredis/pubsub_connection.rb [hunk 10]
```diff
@@ -108,14 +111,23 @@ def handle_response(reply)
...
         else
-          EM::Hireds.logger.error("#{@name} - unrecognised response: #{reply.inspect}")
+          EM::Hiredis.logger.error("#{@name} - unrecognised response: #{reply.inspect}")
         end
       end
     end
   end
-end
\ No newline at end of file
+end
```
Comment (WillSewell) [inline lib/em-hiredis/pubsub_connection.rb]:
```suggestion
end

```

File: lib/em-hiredis/pubsub_connection.rb [hunk 11]
```diff
@@ -108,14 +111,23 @@ def handle_response(reply)
...
         @auth_df = nil
+      elsif @ping_df
+        if reply.kind_of?(Exception)
+          e = EM::Hiredis::RedisError.new(reply.message)
+          e.redis_error = reply
+          @ping_df.fail(e)
+        else
+          @ping_df.succeed(reply)
+        end
+        @ping_df = nil
```
Comment (WillSewell) [inline lib/em-hiredis/pubsub_connection.rb]: This concerns me because it relies on the user of this class to call "auth" and then "ping" in a specific order. If auth ping is called first, followed by auth, then it'll be the pong response which actually succeeds the `@auth_df`.

IMO this would be much easier to reason about if the same approach was used as the `RedisConnection` module, i.e. to have a `@response_queue` that are processed in the order the commands were sent.

File: spec/client_server_spec.rb [hunk 12]
```diff
@@ -2,6 +2,13 @@
 
 describe EM::Hiredis::Client do
 
+  around(:each) do |test|
+    EM.run do
+      test.run
+      EM.stop
+    end
+  end
```
Comment (WillSewell) [inline spec/client_server_spec.rb]: It's not clear to me why these changes require this to be added when it wasn't here previously. Could you explain this please?

**Comment by WillSewell** [comment]
There are still tests failing. Do you know why this is?

File: spec/client_server_spec.rb (hunk 12 above)
Comment (annzenkina) [reply spec/client_server_spec.rb]: Because in this implementation I rely on EM timeout function. So if there is no EM in the test it gives you an error:
```RuntimeError:
       eventmachine not initialized: evma_install_oneshot_timer```

**Comment by annzenkina** [comment]
I don't know why as these tests also failing in master. We agreed with @leesio that it's a different task to fix them.

File: spec/client_server_spec.rb (hunk 12 above)
Comment (WillSewell) [reply spec/client_server_spec.rb]: You can use https://github.com/joshbuddy/em-spec to run tests in an EventMachine context. So in this case you could replace `mock_connections` with

```ruby
  def mock_connections(expected_connections, uri = 'redis://localhost:6379')
    em {

      connectionEM = EM::Hiredis::MockConnectionEM.new(expected_connections, PubsubTestConnection)

      yield EM::Hiredis::PubsubClient.new(uri, nil, nil, connectionEM), connectionEM.connections

      connectionEM.connections.each { |c| c._expectations_met! }

      done
    }
  end
```

I notice that this should also be done for `client_conn_spec.rb` too. The tests for that are currently failing with `RuntimeError: eventmachine not initialized: evma_install_oneshot_timer`.

You have also added

```ruby
  around(:each) do |test|
    EM.run do
      test.run
      EM.stop
    end
  end
```

to `spec/client_server_spec.rb` and `spec/live/redis_commands_spec.rb`, but these are not necessary since these tests are already using `em` where necessary.

Comment (annzenkina) [reply spec/client_server_spec.rb]: Thanks! I didn't think of changing `mock_connections`, makes sense.

For this one: `client_conn_spec.rb` i deleted EM today because I've confused this test file with the different one. Returned them back.

For this one `spec/client_server_spec.rb` – it works for me either with `127.0.0.1` instead of `localhost` or by adding
```  around(:each) do |test|
    EM.run do
      test.run
      EM.stop
    end
  end```
Not sure I understand why. I will look into it.

**Comment by WillSewell** [comment]
Looks good!

Should be squashed before merge..
